apple/swift-docc
# Add logic to look for overload groups when emitting warnings

Add logic to look for overload groups when emitting warnings and suggestions about ambiguous symbols. If exactly one of the ambiguous symbols is an overload group, then suggest the author simply remove the incorrect disambiguation instead of selecting a different disambiguation hash.

rdar://126382989

## Summary

Today when an author links to a page using an invalid disambiguation hash, DocC emits a warning similar to this listing all of the valid disambiguation hashes:

```
warning: '7w26t' isn't a disambiguation for 'init(content:)' at '/MyFramework/SomeStruct'
 --> SomeStruct.md:7:19-7:25
5 | ### Some Interesting Topic
6 |
7 + - ``init(content:)-7w26t``
  |                   ├─suggestion: Replace '7w26t' with '25rvu' for'something'
  |                   ├─suggestion: Replace '7w26t' with '4tm5u' for'something-else'
  |                   ╰─suggestion: Replace '7w26t' with '5di5z' for'something-even-different'
8 | - ``init(content:)-8mfe``
```

However, we expect that most of the time authors need to specify disambiguation hashes when they are linking directly to one member of a set of overloaded symbols. (Symbols that have the same parameters but different types).

Therefore, this PR changes the diagnostic logic for ambiguous symbols to emit a warning similar to this instead:

```
warning: '7w26t' isn't a disambiguation for 'init(content:)' at '/MyFramework/SomeStruct'
 --> SomeStruct.md:7:19-7:25
5 | ### Some Interesting Topic
6 |
7 + - ``init(content:)-7w26t``
  |                   ╰─suggestion: Remove '-7w26t' to use overload group 'something'
8 | - ``init(content:)-8mfe``
```

Now DocC will encourage writers to use simpler paths when possible, allowing readers to navigate to overload group pages.

We might consider adding logic in a follow-on PR to emit warnings for _valid links_ that use disambiguation hashes unnecessarily.

## Dependencies

None

## Testing

Steps:
1. Compile documentation that contains a broken link due to an incorrect disambiguation hash. Use the `--enable-experimental-overloaded-symbol-presentation` command line argument to enable the experimental overloads feature flag.
2. Check that the target page is one symbol is a set of overloaded symbols
3. Test that DocC emits a warning similar to the example above.
4. Repeat the test when the broken link selects ambiguous symbols that are not all overloads in the same group. DocC should emit the original warning.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x ] Added tests
- [x ] Ran the `./bin/test` script and it succeeded
